lib/filterx/func-str.c

```diff
@@ -497,3 +519,198 @@ filterx_function_includes_new(FilterXFunctionArgs *args, GError **error)
+      if (!filterx_object_extract_string_ref(b_obj, &b_str, &b_str_len))
+        {
+          filterx_eval_push_error("failed to extract string value", self->b.expr, b_obj);
+          filterx_object_unref(b_obj);
+          goto exit;
+        }
+    }
+  else
+    b_str = self->b.literal->str;
+
+  /* TODO: utf-8 support together with all func-str-transform */
+  gint cmp = g_ascii_strcasecmp(a_str, b_str);
+  result = filterx_integer_new(cmp);
+
+  filterx_object_unref(b_obj);
+
+exit:
+  filterx_object_unref(a_obj);
+  return result;
+}
+
+static inline GString *
+_extract_literal(FilterXExpr *expr)
+{
+  FilterXObject *literal = filterx_expr_eval(expr);
+  g_assert(literal);
+
+  const gchar *str = NULL;
+  gsize str_len = 0;
+  if (!filterx_object_extract_string_ref(literal, &str, &str_len))
+    return NULL;
+
+  return g_string_new_len(str, str_len);
+}
+
+static FilterXExpr *
+_strcasecmp_optimize(FilterXExpr *s)
+{
+  FilterXStrcasecmp *self = (FilterXStrcasecmp *) s;
+
+  self->a.expr = filterx_expr_optimize(self->a.expr);
+  self->b.expr = filterx_expr_optimize(self->b.expr);
+
+  if (filterx_expr_is_literal(self->a.expr) && filterx_expr_is_literal(self->b.expr))
+    {
+      FilterXObject *result = _strcasecmp_eval(s);
+      if (!result)
+        goto exit;
+
+      return filterx_literal_new(result);
+    }
+
+  if (filterx_expr_is_literal(self->a.expr))
+    {
+      GString *literal = _extract_literal(self->a.expr);
+      if (!literal)
+        goto exit;
+
+      self->a.literal = literal;
+      self->a_literal = TRUE;
+      goto exit;
+    }
+
+  if (filterx_expr_is_literal(self->b.expr))
+    {
+      GString *literal = _extract_literal(self->b.expr);
+      if (!literal)
+        goto exit;
+
+      self->b.literal = literal;
+      self->b_literal = TRUE;
+    }
+
+exit:
+  return filterx_function_optimize_method(&self->super);
+}
+
+static gboolean
+_strcasecmp_init(FilterXExpr *s, GlobalConfig *cfg)
+{
+  FilterXStrcasecmp *self = (FilterXStrcasecmp *) s;
+
+  if (!self->a_literal && !filterx_expr_init(self->a.expr, cfg))
+    return FALSE;
+
+  if (!self->b_literal && !filterx_expr_init(self->b.expr, cfg))
+    {
+      filterx_expr_deinit(self->a.expr, cfg);
+      return FALSE;
+    }
+
+  return filterx_function_init_method(&self->super, cfg);
+}
+
+static void
+_strcasecmp_deinit(FilterXExpr *s, GlobalConfig *cfg)
+{
+  FilterXStrcasecmp *self = (FilterXStrcasecmp *) s;
+
+  if (!self->a_literal)
+    filterx_expr_deinit(self->a.expr, cfg);
+  if (!self->b_literal)
+    filterx_expr_deinit(self->b.expr, cfg);
+  filterx_function_deinit_method(&self->super, cfg);
+}
+
+static void
+_strcasecmp_free(FilterXExpr *s)
+{
+  FilterXStrcasecmp *self = (FilterXStrcasecmp *) s;
+  if (self->a_literal)
+    g_string_free(self->a.literal, TRUE);
+  else
+    filterx_expr_unref(self->a.expr);
+
+  if (self->b_literal)
+    g_string_free(self->b.literal, TRUE);
+  else
+    filterx_expr_unref(self->b.expr);
```

Comment:
We don't release the expr in optimize, so I think we should free the expr unconditionally here, or do that in optimize.

Reply:
that seems to be in a union, so the expr pointer is not available anymore.